### PR TITLE
Fixed smalldatetime and datetimeoffset datatypes handling

### PR DESCRIPTION
**Pull request by Aryan-Verma**
- Current handling incorrectly maps SmallDateTime datatype to timestamp in the sink side as it does not have any time zone information. Fixed it to map to datetime.
- DateTimeOffset was incorrectly mapped to datetime in the sink side. Fixed it to map to timestamp. Example: A DateTimeOffset value of '2025-12-10 12:32:10 +01:00' is transferred as '2025-12-10 11:32:10 +00:00' to the sink side.

### CODE REVIEW

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java [hunk 1]
```diff
@@ -35,17 +35,29 @@ public boolean isFieldCompatible(Schema.Field field, ResultSetMetaData metadata,
...
     boolean isSigned = metadata.isSigned(index);
     int precision = metadata.getPrecision(index);
 
     // Handles datetime datatypes
     // Case when Timestamp maps to datetime
     // Case when Datetime2 maps to datetime
-    // Case when DatetimeOffset maps to datetime
-    if ((sqlType == Types.TIMESTAMP || sqlType == SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE)
-            && fieldLogicalType.equals(Schema.LogicalType.DATETIME)) {
+    if (sqlType == Types.TIMESTAMP && fieldLogicalType.equals(Schema.LogicalType.DATETIME)) {
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java]: This change is not backward compatible. We should not break the compatibility of DateTimeOffset with DateTime type at the sink as it might regress any existing use case.

**Comment by MrRahulSharma** [comment]
MSSql Source plugin should make sure that the backward compatibility is retained for SmallDateTime and DateTimeOffset.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java (hunk 1 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java]: Added the fix to ensure backward compatibility

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java [hunk 2]
```diff
@@ -203,9 +203,11 @@ protected void validateField(FailureCollector collector, Schema.Field field, Sch
...
       // But we don't have the original source type information here and don't want to do big refactoring here
       if (actualFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME &&
-            expectedFieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS ||
-            actualFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME &&
-              expectedFieldSchema.getType() == Schema.Type.STRING) {
+        expectedFieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS ||
+        actualFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME &&
+          expectedFieldSchema.getType() == Schema.Type.STRING ||
+        actualFieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS &&
+          expectedFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME) {
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java]: Create a new if block and add comments around to state what use case we are trying to cover.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java [hunk 3]
```diff
@@ -67,6 +74,34 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
...
+      case SqlServerSourceSchemaReader.SMALLDATETIME_TYPE:
+        GregorianCalendar gc = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        gc.setGregorianChange(new Date(Long.MIN_VALUE));
+        Timestamp timestampSmalldatetime = resultSet.getTimestamp(columnIndex, gc);
+        if (timestampSmalldatetime != null) {
+          ZonedDateTime zonedDateTime = OffsetDateTime.of(timestampSmalldatetime.toLocalDateTime(),
+                                                          OffsetDateTime.now().getOffset())
+                                                          .atZoneSameInstant(ZoneId.of("UTC"));
+          Schema nonNullableSchema = field.getSchema().isNullable() ?
+            field.getSchema().getNonNullable() : field.getSchema();
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: you can use fieldSchema here instead of creating another variable nonNullableSchema

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java [hunk 4]
```diff
@@ -67,6 +74,34 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
...
+          recordBuilder.set(field.getName(), null);
+        }
+        break;
+      case SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE:
+        OffsetDateTime timestampOffset = resultSet.getObject(columnIndex, OffsetDateTime.class);
+        if (timestampOffset != null) {
+          if (fieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS) {
+            recordBuilder.setTimestamp(field.getName(), timestampOffset.atZoneSameInstant(ZoneId.of("UTC")));
+          } else {
+            recordBuilder.set(field.getName(), resultSet.getString(columnIndex));
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]:
```suggestion
            // Deprecated use case of supporting DateTimeOffset to CDAP DateTime conversion.
            recordBuilder.set(field.getName(), resultSet.getString(columnIndex));
```

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java [hunk 5]
```diff
@@ -67,6 +74,34 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
...
+                                             zonedDateTime);
+        } else {
+          recordBuilder.set(field.getName(), null);
+        }
+        break;
+      case SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE:
+        OffsetDateTime timestampOffset = resultSet.getObject(columnIndex, OffsetDateTime.class);
+        if (timestampOffset != null) {
+          if (fieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS) {
+            recordBuilder.setTimestamp(field.getName(), timestampOffset.atZoneSameInstant(ZoneId.of("UTC")));
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]:
```suggestion
            // DateTimeOffset to CDAP Timestamp type conversion
            recordBuilder.setTimestamp(field.getName(), timestampOffset.atZoneSameInstant(ZoneId.of("UTC")));
```

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java [hunk 6]
```diff
@@ -75,14 +110,22 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
...
         break;
-      case SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE:
-        recordBuilder.set(field.getName(), resultSet.getString(columnIndex));
-        break;
       default:
         setField(resultSet, recordBuilder, field, columnIndex, sqlType, sqlPrecision, sqlScale);
     }
   }
 
+  private void setZonedDateTimeBasedOnOuputSchema(StructuredRecord.Builder recordBuilder,
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: We can move this method's content into the caller block and update the comment about which flow is deprecated.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java [hunk 7]
```diff
@@ -71,12 +75,23 @@ public static boolean shouldConvertToDatetime(ResultSetMetaData metadata, int in
     // datetime and datetime2 will have type Types.TIMESTAMP
     // cannot decide based on sql type
     String columnTypeName = metadata.getColumnTypeName(index);
-    return shouldConvertToDatetime(columnTypeName);
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java:74]: Remove the function shouldConvertToDatetime(String) and move the logic to the default of the switch block.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java [hunk 8]
```diff
@@ -32,6 +32,7 @@ public class SqlServerSourceSchemaReader extends CommonSchemaReader {
   public static final int GEOMETRY_TYPE = -157;
   public static final int GEOGRAPHY_TYPE = -158;
   public static final int SQL_VARIANT = -156;
+  public static final int SMALLDATETIME_TYPE = 93;
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: This is not required, you can directly user java.sql.Types.TIMESTAMP for matching

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java (hunk 2 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java]: Created different if blocks along with their comments

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java (hunk 3 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: removed nonNullableSchema

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java (hunk 4 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java (hunk 5 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java (hunk 6 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java (hunk 7 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java:74]: removed

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java (hunk 8 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java [hunk 9]
```diff
@@ -61,27 +65,29 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
...
+      case Types.TIMESTAMP:
+        // SmallDateTime will also be mapped to Types.DATETIME
+        return true;
+      case DATETIME_OFFSET_TYPE:
+        // DateTimeOffset should map to Types.TIMESTAMP_MICROS
+        return false;
+      default:
+        // DateTime and DateTime2 will be converted to CDAP DateTime
+        return columnTypeName.startsWith(DATETIME_TYPE_PREFIX);
+    }
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: You can refactor this, to check only for Types.TIMESTAMP and default case.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java [hunk 10]
```diff
@@ -53,6 +54,9 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
       return Schema.of(Schema.LogicalType.DATETIME);
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: Nit : Remove empty line.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java [hunk 11]
```diff
@@ -35,17 +35,30 @@ public boolean isFieldCompatible(Schema.Field field, ResultSetMetaData metadata,
...
+    if (sqlType == SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE &&
+      fieldLogicalType.equals(Schema.LogicalType.TIMESTAMP_MICROS)) {
+      return true;
+    }
+
+    // Case when smalldatetime maps to datetime
+    if (sqlTypeName.equalsIgnoreCase("smalldatetime")
+      && fieldLogicalType.equals(Schema.LogicalType.DATETIME)) {
+      return true;
+    }
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java]: Why is this required, If I understand correctly the above block at line 45, should already cover for this use case.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java [hunk 12]
```diff
@@ -152,7 +153,16 @@ protected String getStratifiedQuery(String tableName, int limit, String strata,
   @Override
   protected Schema getSchema(int sqlType, String typeName, int scale, int precision, String columnName,
                              boolean isSigned, boolean handleAsDecimal) throws SQLException {
-    if (SqlServerSourceSchemaReader.shouldConvertToDatetime(typeName)) {
+    if (Types.TIMESTAMP == sqlType) {
+      // SmallDateTime case
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: This should ideally cover Datetime, Datetime2 and SmallDatetime all 3 of them.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java [hunk 13]
```diff
@@ -152,7 +153,16 @@ protected String getStratifiedQuery(String tableName, int limit, String strata,
...
+      return Schema.of(Schema.LogicalType.DATETIME);
+    }
+
+    if (SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE == sqlType) {
+      return Schema.of(Schema.LogicalType.TIMESTAMP_MICROS);
+    }
+
+    if (typeName.startsWith(SqlServerSourceSchemaReader.DATETIME_TYPE_PREFIX)) {
       return Schema.of(Schema.LogicalType.DATETIME);
     }
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: This block may not be required.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java (hunk 9 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java (hunk 10 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java (hunk 11 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java]: removed this block

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java (hunk 12 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java (hunk 13 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: removed this

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java [hunk 14]
```diff
@@ -35,17 +35,24 @@ public boolean isFieldCompatible(Schema.Field field, ResultSetMetaData metadata,
     Schema.LogicalType fieldLogicalType = fieldSchema.getLogicalType();
 
     int sqlType = metadata.getColumnType(index);
+    String sqlTypeName = metadata.getColumnTypeName(index);
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java]: This is not required.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSinkDBRecord.java [hunk 15]
```diff
@@ -55,7 +57,18 @@ protected void writeNonNullToDB(PreparedStatement stmt, Schema fieldSchema,
                                   String fieldName, int fieldIndex) throws SQLException {
     Object fieldValue = (fieldName != null) ? record.get(fieldName) : null;
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSinkDBRecord.java]: This can be moved to the switch GEOGRAPHY_TYPE and GEOMETRY_TYPE block below.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java [hunk 16]
```diff
@@ -67,6 +79,36 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
...
+        GregorianCalendar gc = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        gc.setGregorianChange(new Date(Long.MIN_VALUE));
+        Timestamp timestampSmalldatetime = resultSet.getTimestamp(columnIndex, gc);
+        if (timestampSmalldatetime != null) {
+          ZonedDateTime zonedDateTime = OffsetDateTime.of(timestampSmalldatetime.toLocalDateTime(),
+                                                          OffsetDateTime.now().getOffset())
+                                                          .atZoneSameInstant(ZoneId.of("UTC"));
+          if (Schema.LogicalType.DATETIME.equals(logicalType)) {
+            // SmallDateTime to CDAP DateTime type conversion
+            recordBuilder.setDateTime(field.getName(), zonedDateTime.toLocalDateTime());
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: Why do we need this code ?

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java (hunk 14 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java]: removed

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSinkDBRecord.java (hunk 15 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSinkDBRecord.java]: done

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java (hunk 16 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: removed this. Didn't notice that this case was already covered above

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java [hunk 17]
```diff
@@ -57,9 +63,9 @@ public boolean isFieldCompatible(Schema.Field field, ResultSetMetaData metadata,
           || sqlType == SqlServerSinkSchemaReader.GEOMETRY_TYPE
           || super.isFieldCompatible(field, metadata, index);
       case STRING:
-        return sqlType == SqlServerSinkSchemaReader.DATETIME_OFFSET_TYPE
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java:60]: Why are we removing this ?

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java [hunk 18]
```diff
@@ -152,10 +153,15 @@ protected String getStratifiedQuery(String tableName, int limit, String strata,
...
+    if (Types.TIMESTAMP == sqlType) {
+      // DateTime, DateTime2 and SmallDateTime cases
       return Schema.of(Schema.LogicalType.DATETIME);
     }
 
+    if (SqlServerSourceSchemaReader.DATETIME_OFFSET_TYPE == sqlType) {
+      return Schema.of(Schema.LogicalType.TIMESTAMP_MICROS);
+    }
+
     if (SqlServerSourceSchemaReader.GEOMETRY_TYPE == sqlType || SqlServerSourceSchemaReader.GEOGRAPHY_TYPE == sqlType) {
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: How are we validating this ?

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java [hunk 19]
```diff
@@ -203,9 +203,19 @@ protected void validateField(FailureCollector collector, Schema.Field field, Sch
...
-            expectedFieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS ||
-            actualFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME &&
-              expectedFieldSchema.getType() == Schema.Type.STRING) {
+        expectedFieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS) {
+        // SmallDateTime case where we map it to CDAP DateTime now as opposed to CDAP Timestamp earlier, as
+        // SmallDateTime does not contain any TimeZone information
+        return;
+      }
+      if (actualFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME &&
+          expectedFieldSchema.getType() == Schema.Type.STRING) {
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java]:
```suggestion
      if ((actualFieldSchema.getLogicalType() == Schema.LogicalType.DATETIME ||
      actualFieldSchema.getLogicalType() == Schema.LogicalType.TIMESTAMP_MICROS)
      && expectedFieldSchema.getType() == Schema.Type.STRING) {
```

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java [hunk 20]
```diff
@@ -61,27 +64,23 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
...
-   * @return Whether the corresponding data type should be converted to CDAP Datetime Logical Type
-   */
-  public static boolean shouldConvertToDatetime(String typeName) {
-    return typeName.startsWith(DATETIME_TYPE_PREFIX);
+    if (columnSqlType == Types.TIMESTAMP) {
+      // SmallDateTime will also be mapped to Types.DATETIME
+      return true;
+    }
+    // DateTime and DateTime2 will be converted to CDAP DateTime
+    return columnTypeName.startsWith(DATETIME_TYPE_PREFIX);
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]:
```suggestion
    return columnSqlType == Types.TIMESTAMP;
```

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java [hunk 21]
```diff
@@ -61,27 +64,23 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
...
    * @param metadata result set metadata
    * @param index index of the column
    * @return whether the corresponding column should be converted to CDAP Datetime Logical Type
    * @throws SQLException
    */
   public static boolean shouldConvertToDatetime(ResultSetMetaData metadata, int index) throws SQLException {
-    // datetimeoffset will have type DATETIME_OFFSET_TYPE
-    // datetime and datetime2 will have type Types.TIMESTAMP
     // cannot decide based on sql type
     String columnTypeName = metadata.getColumnTypeName(index);
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: This can be removed.

**Comment by MrRahulSharma** [comment]
Also, add E2E tests for this change

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java (hunk 17 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlFieldsValidator.java:60]: added it again

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java (hunk 18 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: This method is not being called anywhere so removed it altogether

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java (hunk 21 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java]: removed the method

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java [hunk 22]
```diff
@@ -148,17 +149,4 @@ protected String getStratifiedQuery(String tableName, int limit, String strata,
...
-                             boolean isSigned, boolean handleAsDecimal) throws SQLException {
-    if (SqlServerSourceSchemaReader.shouldConvertToDatetime(typeName)) {
-      return Schema.of(Schema.LogicalType.DATETIME);
-    }
-
-    if (SqlServerSourceSchemaReader.GEOMETRY_TYPE == sqlType || SqlServerSourceSchemaReader.GEOGRAPHY_TYPE == sqlType) {
-      return Schema.of(Schema.Type.BYTES);
-    }
-    return super.getSchema(sqlType, typeName, scale, precision, columnName, isSigned, handleAsDecimal);
-  }
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: Why are we removing this block ?

Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: It is not called anywhere

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java [hunk 23]
```diff
@@ -40,6 +40,7 @@
 
 import java.io.IOException;
 import java.sql.SQLException;
+import java.sql.Types;
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: Remove this import.

File: mssql-plugin/src/e2e-test/resources/pluginParameters.properties [hunk 24]
```diff
@@ -40,8 +40,8 @@ uniqueIdentifierColumnsList=(ID, COL1)
...
 
-dateTimeColumns=(ID VARCHAR(100) PRIMARY KEY, COL1 DATETIME, COL2 DATETIME2(0))
-dateTimeColumnsList=(ID, COL1, COL2)
-dateTimeValues=VALUES ('User1', '2023-01-01 01:00:00.000', '2023-01-01 01:00:00.000')
+dateTimeColumns=(ID VARCHAR(100) PRIMARY KEY, COL1 DATETIME, COL2 DATETIME2(0), COL3 SMALLDATETIME)
+dateTimeColumnsList=(ID, COL1, COL2, COL3)
+dateTimeValues=VALUES ('User1', '2023-01-01 01:00:00.000', '2023-01-01 01:00:00.000', '2023-01-01 01:00:00.000')
 outputDatatypesSchema4=[{"key":"ID","value":"string"},{"key":"COL1","value":"datetime"},\
-  {"key":"COL2","value":"datetime"}]
+  {"key":"COL2","value":"datetime"},{"key":"COL3","value":"datetime"}]
```
Comment (MrRahulSharma) [inline mssql-plugin/src/e2e-test/resources/pluginParameters.properties]: Please add a test for DatetimeOffset type as well.

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java [hunk 25]
```diff
@@ -52,21 +58,35 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
...
-      } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
-        throw new RuntimeException(String.format("Fail to convert column %s of type %s to datetime. Error: %s.",
-                                                 resultSet.getMetaData().getColumnName(columnIndex),
-                                                 resultSet.getMetaData().getColumnTypeName(columnIndex),
-                                                 e.getMessage()), e);
-      }
-    }
     switch (sqlType) {
+      case Types.TIMESTAMP:
+        // SmallDateTime case
```
Comment (MrRahulSharma) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: Correction 
```suggestion
        // SmallDateTime, Datetime, Datetime2 usecase
```

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java (hunk 23 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerConnector.java]: removed

File: mssql-plugin/src/e2e-test/resources/pluginParameters.properties (hunk 24 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/e2e-test/resources/pluginParameters.properties]: added

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java (hunk 25 above)
Comment (Aryan-Verma) [reply mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceDBRecord.java]: done

**Review by MrRahulSharma — APPROVED** [review]
LGTM, Please squash the commits before merging.
